From 320281d8267943efe200b20f2106fab53cdee123 Mon Sep 17 00:00:00 2001
From: Ben Walker <ben@nvidia.com>
Date: Mon, 16 Sep 2024 13:17:03 -0700
Subject: [PATCH 12/17] nvmf/tcp: Use zero copy recv APIs efficiently for
 in-capsule buffers

When using the zero copy receive API for sockets, the sockets simply
hand us the next segment of the TCP stream that the socket layer has
already allocated. This means we no longer need to allocate in-capsule
buffers of our own - we can simply make the requests point at this
memory and use reference counts to free it at the correct time.

This results in a major simplification, removing the in-capsule buffer
array from each request and the control message pool for admin and
fabrics commands. This also allows us to support any size
in-capsule data with no memory overhead. However, we leave the default
sizes in place for now.

Unfortunately, zero copy receive with POSIX sockets is much less
efficient than the old code path. So all of the old code path remains in
place and can't be deleted.

Change-Id: I88befb569a49075d6dbdca12f2c3e1df8498e30a
Signed-off-by: Ben Walker <ben@nvidia.com>
---
 lib/nvmf/tcp.c                    | 213 ++++++++++++++++++++++++++++--
 test/unit/lib/nvmf/tcp.c/tcp_ut.c |   3 -
 2 files changed, 201 insertions(+), 15 deletions(-)

diff --git a/lib/nvmf/tcp.c b/lib/nvmf/tcp.c
index e4e7be672..72e586b02 100644
--- a/lib/nvmf/tcp.c
+++ b/lib/nvmf/tcp.c
@@ -262,6 +262,9 @@ struct spdk_nvmf_tcp_req  {
 	STAILQ_ENTRY(spdk_nvmf_tcp_req)		link;
 	TAILQ_ENTRY(spdk_nvmf_tcp_req)		state_link;
 	STAILQ_ENTRY(spdk_nvmf_tcp_req)		control_msg_link;
+
+	struct spdk_nvmf_tcp_stream_segment	**segment_refs;
+	uint32_t				segment_cnt;
 };
 
 /* This tracks a segment of the TCP stream */
@@ -269,6 +272,7 @@ struct spdk_nvmf_tcp_stream_segment {
 	void						*buf;
 	uint32_t					len;
 	struct spdk_sock_buf_token			*token;
+	uint32_t					refcnt;
 	STAILQ_ENTRY(spdk_nvmf_tcp_stream_segment)	link;
 };
 
@@ -312,6 +316,7 @@ struct spdk_nvmf_tcp_qpair {
 	void					*bufs;
 	struct spdk_nvmf_tcp_req		*reqs;
 	struct nvme_tcp_pdu			*pdus;
+	struct spdk_nvmf_tcp_stream_segment	**segment_refs;
 	uint32_t				resource_count;
 	uint32_t				recv_buf_size;
 
@@ -639,6 +644,12 @@ static void
 nvmf_tcp_segment_put(struct spdk_nvmf_tcp_qpair *tqpair,
 		     struct spdk_nvmf_tcp_stream_segment *segment)
 {
+	assert(segment->refcnt > 0);
+	segment->refcnt--;
+	if (segment->refcnt != 0) {
+		return;
+	}
+
 	if (tqpair->group != NULL) {
 		/* Release the buffer through the poll group's sock_group rather than the socket.
 		 * The socket may already be closed during qpair teardown, but the buffer pool is
@@ -710,6 +721,7 @@ _nvmf_tcp_qpair_destroy(void *_tqpair)
 	 */
 	spdk_poller_unregister(&tqpair->timeout_poller);
 	spdk_dma_free(tqpair->pdus);
+	free(tqpair->segment_refs);
 	free(tqpair->reqs);
 	spdk_free(tqpair->bufs);
 	spdk_trace_unregister_owner(tqpair->qpair.trace_id);
@@ -1412,7 +1424,7 @@ nvmf_tcp_qpair_init_mem_resource(struct spdk_nvmf_tcp_qpair *tqpair)
 		return -1;
 	}
 
-	if (in_capsule_data_size) {
+	if (in_capsule_data_size && !tqpair->use_zero_copy_rx) {
 		tqpair->bufs = spdk_zmalloc(tqpair->resource_count * in_capsule_data_size, 0x1000,
 					    NULL, SPDK_ENV_LCORE_ID_ANY,
 					    SPDK_MALLOC_DMA);
@@ -1421,6 +1433,16 @@ nvmf_tcp_qpair_init_mem_resource(struct spdk_nvmf_tcp_qpair *tqpair)
 			return -1;
 		}
 	}
+
+	if (tqpair->use_zero_copy_rx) {
+		tqpair->segment_refs = calloc(tqpair->resource_count * NVMF_REQ_MAX_BUFFERS,
+					      sizeof(*tqpair->segment_refs));
+		if (tqpair->segment_refs == NULL) {
+			SPDK_ERRLOG("Unable to allocate segment refs on tqpair=%p.\n", tqpair);
+			return -1;
+		}
+	}
+
 	/* prepare memory space for receiving pdus and tcp_req */
 	/* Add additional 1 member, which will be used for mgmt_pdu owned by the tqpair */
 	tqpair->pdus = spdk_dma_zmalloc((2 * tqpair->resource_count + 1) * sizeof(*tqpair->pdus), 0x1000,
@@ -1443,6 +1465,9 @@ nvmf_tcp_qpair_init_mem_resource(struct spdk_nvmf_tcp_qpair *tqpair)
 		if (tqpair->bufs) {
 			tcp_req->buf = (void *)((uintptr_t)tqpair->bufs + (i * in_capsule_data_size));
 		}
+		if (tqpair->segment_refs) {
+			tcp_req->segment_refs = &tqpair->segment_refs[i * NVMF_REQ_MAX_BUFFERS];
+		}
 
 		/* Set the cmdn and rsp */
 		tcp_req->req.rsp = (union nvmf_c2h_msg *)&tcp_req->rsp;
@@ -2600,6 +2625,125 @@ err:
 	nvmf_tcp_send_c2h_term_req(tqpair, pdu, fes, error_offset);
 }
 
+static int
+nvmf_tcp_double_buffer_scattered_payload(struct spdk_nvmf_tcp_qpair *tqpair,
+		struct spdk_nvmf_tcp_req *tcp_req, struct nvme_tcp_pdu *pdu)
+{
+	struct spdk_nvmf_tcp_stream_segment *segment;
+	size_t total_sz = 0;
+	void *double_buf;
+	void *buf;
+	size_t len;
+	int i, j;
+	size_t copy_offset;
+
+	for (i = 0; i < (int)pdu->data_iovcnt; i++) {
+		total_sz += pdu->data_iov[i].iov_len;
+	}
+
+	double_buf = spdk_malloc(total_sz, 4096, NULL, SPDK_ENV_SOCKET_ID_ANY, 0);
+	if (double_buf == NULL) {
+		SPDK_ERRLOG("Failed to allocate buffer for scattered payload\n");
+		return -ENOMEM;
+	}
+	assert(tcp_req->segment_refs != NULL);
+
+	/* We walk backward as we copy because we may not get all the way to the beginning of the tcp_req iov array. */
+	copy_offset = total_sz;
+	for (i = (int)pdu->data_iovcnt - 1, j = tcp_req->req.iovcnt - 1; i > -1; i--, j--) {
+		buf = pdu->data_iov[i].iov_base;
+		len = pdu->data_iov[i].iov_len;
+		segment = tcp_req->segment_refs[j];
+
+		tcp_req->req.iov[j].iov_base = NULL;
+		tcp_req->req.iov[j].iov_len = 0;
+		tcp_req->req.iovcnt--;
+
+		tcp_req->segment_refs[j] = NULL;
+		tcp_req->segment_cnt--;
+
+		pdu->data_iov[i].iov_base = NULL;
+		pdu->data_iov[i].iov_len = 0;
+		pdu->data_iovcnt--;
+
+		copy_offset -= len;
+		memcpy((uint8_t *)double_buf + copy_offset, buf, len);
+
+		nvmf_tcp_segment_put(tqpair, segment);
+	}
+
+	tcp_req->req.iov[tcp_req->req.iovcnt].iov_base = double_buf;
+	tcp_req->req.iov[tcp_req->req.iovcnt].iov_len = total_sz;
+	tcp_req->req.iovcnt++;
+
+	/* A null entry is a signal that this was double buffered. */
+	tcp_req->segment_refs[tcp_req->segment_cnt] = NULL;
+	tcp_req->segment_cnt++;
+
+	return 0;
+}
+
+static int
+nvmf_tcp_read_data_zcopy(struct spdk_nvmf_tcp_qpair *tqpair, struct spdk_nvmf_tcp_req *tcp_req,
+			 struct nvme_tcp_pdu *pdu)
+{
+	struct spdk_nvmf_tcp_stream_segment *segment;
+	uint32_t length;
+	int ret;
+
+	length = pdu->data_len - pdu->rw_offset;
+	assert(tcp_req->segment_refs != NULL);
+
+	ret = 0;
+	while (length > 0) {
+		int len;
+
+		segment = STAILQ_FIRST(&tqpair->pending_stream);
+		if (segment == NULL) {
+			break;
+		}
+
+		assert(segment->len > tqpair->stream_offset);
+
+		len = spdk_min(segment->len - tqpair->stream_offset, (uint32_t)length);
+
+		if (pdu->data_iovcnt >= SPDK_COUNTOF(pdu->data_iov) - 1 ||
+		    tcp_req->req.iovcnt >= SPDK_COUNTOF(tcp_req->req.iov) - 1) {
+			ret = nvmf_tcp_double_buffer_scattered_payload(tqpair, tcp_req, pdu);
+			if (ret != 0) {
+				return ret;
+			}
+		}
+
+		pdu->data_iov[pdu->data_iovcnt].iov_base = segment->buf + tqpair->stream_offset;
+		pdu->data_iov[pdu->data_iovcnt].iov_len = len;
+		pdu->data_iovcnt++;
+
+		tcp_req->req.iov[tcp_req->req.iovcnt].iov_base = segment->buf + tqpair->stream_offset;
+		tcp_req->req.iov[tcp_req->req.iovcnt].iov_len = len;
+		tcp_req->req.iovcnt++;
+
+		tcp_req->segment_refs[tcp_req->segment_cnt] = segment;
+		tcp_req->segment_cnt++;
+		segment->refcnt++;
+
+		tqpair->stream_offset += len;
+		length -= len;
+		ret += len;
+
+		if (tqpair->stream_offset == segment->len) {
+			tqpair->stream_offset = 0;
+			STAILQ_REMOVE_HEAD(&tqpair->pending_stream, link);
+
+			assert(segment->refcnt > 0);
+			segment->refcnt--;
+			assert(segment->refcnt != 0);
+		}
+	}
+
+	return ret;
+}
+
 static int
 nvmf_tcp_read_data(struct spdk_nvmf_tcp_qpair *tqpair, int bytes,
 		   void *_buf)
@@ -2665,6 +2809,7 @@ static int
 nvmf_tcp_read_payload_data(struct spdk_nvmf_tcp_qpair *tqpair, struct nvme_tcp_pdu *pdu)
 {
 	struct iovec iov[NVME_TCP_MAX_SGL_DESCRIPTORS + 1];
+	struct spdk_nvmf_tcp_req *tcp_req;
 	int iovcnt;
 	struct spdk_iov_sgl sgl;
 	int i;
@@ -2673,6 +2818,12 @@ nvmf_tcp_read_payload_data(struct spdk_nvmf_tcp_qpair *tqpair, struct nvme_tcp_p
 	assert(tqpair != NULL);
 	assert(tqpair->sock != NULL);
 
+	tcp_req = pdu->req;
+	if (tqpair->use_zero_copy_rx && tcp_req != NULL && tcp_req->has_in_capsule_data) {
+		/* We'll do a zero copy recv into the PDU. */
+		return nvmf_tcp_read_data_zcopy(tqpair, tcp_req, pdu);
+	}
+
 	spdk_iov_sgl_init(&sgl, iov, NVME_TCP_MAX_SGL_DESCRIPTORS + 1, pdu->rw_offset);
 
 	for (i = 0; i < (int)pdu->data_iovcnt; i++) {
@@ -2973,7 +3124,12 @@ nvmf_tcp_req_parse_sgl(struct spdk_nvmf_tcp_req *tcp_req,
 			goto fatal_err;
 		}
 
-		if (spdk_unlikely(length > max_len)) {
+		if (tqpair->use_zero_copy_rx) {
+			/* We don't fill out req->iov. It will be filled in as data arrives. */
+			req->iovcnt = 0;
+			assert(tcp_req->segment_refs != NULL);
+			tcp_req->segment_cnt = 0;
+		} else if (length > max_len) {
 			/* According to the SPEC we should support ICD up to 8192 bytes for admin and fabric commands */
 			if (length <= SPDK_NVME_TCP_IN_CAPSULE_DATA_MAX_SIZE &&
 			    (cmd->opc == SPDK_NVME_OPC_FABRIC || req->qpair->qid == 0)) {
@@ -2989,6 +3145,8 @@ nvmf_tcp_req_parse_sgl(struct spdk_nvmf_tcp_req *tcp_req,
 					nvmf_tcp_qpair_set_recv_state(tqpair, NVME_TCP_PDU_RECV_STATE_AWAIT_PDU_BUF);
 					return;
 				}
+				req->iov[0].iov_len = length;
+				req->iovcnt = 1;
 			} else {
 				SPDK_ERRLOG("In-capsule data length 0x%x exceeds capsule length 0x%x\n",
 					    length, max_len);
@@ -2997,6 +3155,8 @@ nvmf_tcp_req_parse_sgl(struct spdk_nvmf_tcp_req *tcp_req,
 			}
 		} else {
 			req->iov[0].iov_base = tcp_req->buf;
+			req->iov[0].iov_len = length;
+			req->iovcnt = 1;
 		}
 
 		req->length = length;
@@ -3284,7 +3444,7 @@ nvmf_tcp_req_process(struct spdk_nvmf_tcp_transport *ttransport,
 				break;
 			}
 
-			assert(tcp_req->req.iovcnt > 0);
+			assert(tqpair->use_zero_copy_rx || tcp_req->req.iovcnt > 0);
 
 			/* If data is transferring from host to controller, we need to do a transfer from the host. */
 			if (tcp_req->req.xfer == SPDK_NVME_DATA_HOST_TO_CONTROLLER) {
@@ -3300,8 +3460,13 @@ nvmf_tcp_req_process(struct spdk_nvmf_tcp_transport *ttransport,
 					SPDK_DEBUGLOG(nvmf_tcp, "Not need to send r2t for tcp_req(%p) on tqpair=%p\n", tcp_req,
 						      tqpair);
 					/* No need to send r2t, contained in the capsuled data */
-					nvme_tcp_pdu_set_data_buf(pdu, tcp_req->req.iov, tcp_req->req.iovcnt,
-								  0, tcp_req->req.length);
+					if (tqpair->use_zero_copy_rx) {
+						pdu->data_len = tcp_req->req.length;
+						pdu->data_iovcnt = 0;
+					} else {
+						nvme_tcp_pdu_set_data_buf(pdu, tcp_req->req.iov, tcp_req->req.iovcnt,
+									  0, tcp_req->req.length);
+					}
 					nvmf_tcp_qpair_set_recv_state(tqpair, NVME_TCP_PDU_RECV_STATE_AWAIT_PDU_PAYLOAD);
 				}
 				break;
@@ -3455,14 +3620,35 @@ nvmf_tcp_req_process(struct spdk_nvmf_tcp_transport *ttransport,
 
 			if (tcp_req->req.data_from_pool) {
 				spdk_nvmf_request_free_buffers(&tcp_req->req, group, transport);
-			} else if (spdk_unlikely(tcp_req->has_in_capsule_data &&
-						 (tcp_req->cmd.opc == SPDK_NVME_OPC_FABRIC ||
-						  tqpair->qpair.qid == 0) && tcp_req->req.length > transport->opts.in_capsule_data_size)) {
+			} else if (tcp_req->has_in_capsule_data) {
 				tgroup = SPDK_CONTAINEROF(group, struct spdk_nvmf_tcp_poll_group, group);
-				assert(tgroup->control_msg_list);
-				SPDK_DEBUGLOG(nvmf_tcp, "Put buf to control msg list\n");
-				nvmf_tcp_control_msg_put(tgroup->control_msg_list,
-							 tcp_req->req.iov[0].iov_base);
+
+				if (!tqpair->use_zero_copy_rx) {
+					if (spdk_unlikely((tcp_req->cmd.opc == SPDK_NVME_OPC_FABRIC || tqpair->qpair.qid == 0) &&
+							  tcp_req->req.length > transport->opts.in_capsule_data_size)) {
+						assert(tgroup->control_msg_list);
+						SPDK_DEBUGLOG(nvmf_tcp, "Put buf to control msg list\n");
+						nvmf_tcp_control_msg_put(tgroup->control_msg_list,
+									 tcp_req->req.iov[0].iov_base);
+					}
+				} else {
+					struct spdk_nvmf_tcp_stream_segment *segment;
+					uint32_t i;
+
+					assert(tcp_req->segment_refs != NULL);
+					for (i = 0; i < tcp_req->segment_cnt; i++) {
+						segment = tcp_req->segment_refs[i];
+
+						if (segment == NULL) {
+							/* This signals a double buffer */
+							spdk_free(tcp_req->req.iov[i].iov_base);
+						} else {
+							nvmf_tcp_segment_put(tqpair, segment);
+						}
+					}
+
+					tcp_req->segment_cnt = 0;
+				}
 			} else if (tcp_req->req.zcopy_bdev_io != NULL) {
 				/* If the request has an unreleased zcopy bdev_io, it's either a
 				 * read, a failed write, or the qpair is being disconnected */
@@ -3474,6 +3660,7 @@ nvmf_tcp_req_process(struct spdk_nvmf_tcp_transport *ttransport,
 				spdk_nvmf_request_zcopy_end(&tcp_req->req, false);
 				break;
 			}
+
 			tcp_req->req.length = 0;
 			tcp_req->req.iovcnt = 0;
 			tcp_req->fused_failed = false;
@@ -3534,6 +3721,8 @@ nvmf_tcp_qpair_process(struct spdk_nvmf_tcp_qpair *tqpair)
 			}
 
 			segment->len = rc;
+			assert(segment->refcnt == 0);
+			segment->refcnt = 1;
 
 			STAILQ_REMOVE_HEAD(&tgroup->free_segments, link);
 			STAILQ_INSERT_TAIL(&tqpair->pending_stream, segment, link);
diff --git a/test/unit/lib/nvmf/tcp.c/tcp_ut.c b/test/unit/lib/nvmf/tcp.c/tcp_ut.c
index 79cfa16ab..32bd5124f 100644
--- a/test/unit/lib/nvmf/tcp.c/tcp_ut.c
+++ b/test/unit/lib/nvmf/tcp.c/tcp_ut.c
@@ -880,14 +880,12 @@ test_nvmf_tcp_qpair_init_mem_resource(void)
 	CU_ASSERT(rc == 0);
 	CU_ASSERT(tqpair->resource_count == SPDK_NVMF_TCP_DEFAULT_MAX_IO_QUEUE_DEPTH);
 	CU_ASSERT(tqpair->reqs != NULL);
-	CU_ASSERT(tqpair->bufs != NULL);
 	CU_ASSERT(tqpair->pdus != NULL);
 	/* Just to check the first and last entry */
 	CU_ASSERT(tqpair->reqs[0].ttag == 1);
 	CU_ASSERT(tqpair->reqs[0].req.qpair == &tqpair->qpair);
 	CU_ASSERT(tqpair->reqs[0].pdu == &tqpair->pdus[0]);
 	CU_ASSERT(tqpair->reqs[0].pdu->qpair == &tqpair->qpair);
-	CU_ASSERT(tqpair->reqs[0].buf == (void *)((uintptr_t)tqpair->bufs));
 	CU_ASSERT(tqpair->reqs[0].req.rsp == (void *)&tqpair->reqs[0].rsp);
 	CU_ASSERT(tqpair->reqs[0].req.cmd == (void *)&tqpair->reqs[0].cmd);
 	CU_ASSERT(tqpair->reqs[0].state == TCP_REQUEST_STATE_FREE);
@@ -895,7 +893,6 @@ test_nvmf_tcp_qpair_init_mem_resource(void)
 	CU_ASSERT(tqpair->reqs[127].req.qpair == &tqpair->qpair);
 	CU_ASSERT(tqpair->reqs[127].pdu == &tqpair->pdus[127]);
 	CU_ASSERT(tqpair->reqs[127].pdu->qpair == &tqpair->qpair);
-	CU_ASSERT(tqpair->reqs[127].buf == (void *)((uintptr_t)tqpair->bufs) + 127 * 4096);
 	CU_ASSERT(tqpair->reqs[127].req.rsp == (void *)&tqpair->reqs[127].rsp);
 	CU_ASSERT(tqpair->reqs[127].req.cmd == (void *)&tqpair->reqs[127].cmd);
 	CU_ASSERT(tqpair->reqs[127].state == TCP_REQUEST_STATE_FREE);
